common/fs/src/cache/tailed_file.rs

```diff
@@ -1,109 +1,233 @@
+#![allow(clippy::await_holding_refcell_ref)]
 use http::types::body::LineBuilder;
 use metrics::Metrics;
 
-use futures::Stream;
+use futures::io::AsyncBufRead;
+use futures::task::{Context, Poll};
+use futures::{ready, Stream, StreamExt};
 
-use std::fs::{File, OpenOptions};
-use std::io::{BufRead, BufReader, Seek, SeekFrom};
+use pin_project_lite::pin_project;
+
+use std::cell::RefCell;
+use std::convert::TryInto;
+use std::fs::OpenOptions;
+use std::io;
+use std::mem;
 use std::path::{Path, PathBuf};
+use std::pin::Pin;
+use std::rc::Rc;
+use tokio::io::{BufReader, SeekFrom};
+use tokio_util::compat::{Compat, Tokio02AsyncReadCompatExt};
+
+fn read_until_internal<R: AsyncBufRead + ?Sized>(
```

Comment:
Lifted straight from the futures::io crate

Reply:
Unfortunately not without a fairly significant amount of overhead, as I can't work out a way to feed in a pointer from outside and repeatedly yield it. We would need some unsafe code or maybe an RC<RefCell> for the output in order to avoid allocating.

I agree that it's unfortunate, but it should also never need to change and it's been fairly battle tested by the rest of the ecosystem